sphinx.go

```diff
@@ -660,25 +677,64 @@ func NewRouter(nodeKey *btcec.PrivateKey, net *chaincfg.Params) *Router {
 // In the case of a successful packet processing, and ProcessedPacket struct is
 // returned which houses the newly parsed packet, along with instructions on
 // what to do next.
-func (r *Router) ProcessOnionPacket(onionPkt *OnionPacket, assocData []byte) (*ProcessedPacket, error) {
-	dhKey := onionPkt.EphemeralKey
-	routeInfo := onionPkt.RoutingInfo
-	headerMac := onionPkt.HeaderMAC
+func (r *Router) ProcessOnionPacket(onionPkt *OnionPacket,
+	assocData []byte) (*ProcessedPacket, error) {
 
+	// Compute the shared secret for this onion packet.
 	sharedSecret, err := r.generateSharedSecret(onionPkt.EphemeralKey)
 	if err != nil {
 		return nil, err
 	}
 
-	// In order to mitigate replay attacks, if we've seen this particular
-	// shared secret before, cease processing and just drop this forwarding
-	// message.
-	r.RLock()
-	if _, ok := r.seenSecrets[sharedSecret]; ok {
-		r.RUnlock()
-		return nil, ErrReplayedPacket
+	// Additionally, compute the hash prefix of the shared secret, which
+	// will serve as an identifier for detecting replayed packets.
+	hashPrefix := hashSharedSecret(&sharedSecret)
+
+	// Continue to optimistically process this packet, deferring replay
+	// protection until the end to reduce the penalty of multiple IO
+	// operations.
+	packet, err := processOnionPacket(onionPkt, &sharedSecret, assocData)
+	if err != nil {
+		return nil, err
+	}
+
+	// Retrieve the packet's forwarding instructions, which will contain the
+	// CLTV at which the packet expires.
+	hopData := packet.ForwardingInstructions
+
+	// Atomically compare this hash prefix with the contents of the on-disk
+	// log, persisting it only if this entry was not detected as a replay.
+	if err := r.log.Put(&hashPrefix, hopData.OutgoingCltv); err != nil {
```

Comment:
This should actually be the _incoming_ CLTV. So the CLTV of the HTLC's onion packet that we processed. As is now, we may end up deleting an entry early since: `incoming_cltv < outgoing_cltv`. 

It's arguable that this is fine as is though, as the only downside (I can see atm) is that we hold onto this entry a bit longer than we should.

Reply:
Shouldn't the incoming cltv > outgoing cltv? Otherwise the timelocks would not be decrementing?

I'm having trouble deciphering why the last statement doesn't contradict the first, I think we would actually be deleting entries early

Reply:
I guess we would need to pass in the `Timeout` from the payment descriptor, both in the batched and singular versions